File: ComputerSimulation/Task1__CR_set_localisation/CR_Localizator.py
from collections import defaultdict
from typing import List
from geometry import Homeomorphism, Domain
from graph import SymbolicImageGraph
import networkx as nx

from interface import Interface

from time import time


class CR_Localizator:
    def __init__(self, n_iterations: float):
        self._n_iterations = n_iterations

    def solve(self, f: Homeomorphism, domain: Domain) -> Domain:
        last_domain = None

        iteration = 0
        while iteration < self._n_iterations:
            print(f'===Iteration {iteration}===')
            i = Interface()

            start = time()
            graph = SymbolicImageGraph(f, domain)
            print(f'Construct Symbolic Image Graph {time() - start}')

            print('nodes', len(graph._graph.nodes))

            start = time()
            if not graph.nonreturnable_vertexes_successfully_deleted():
                print('There is no any CR set')
                return None
            print(f'Delete non-returnable vertexes {time() - start}')

            start = time()
            domain.delete_nonreturnable_cells(graph)
            print(f'Delete non-returnable cellls {time() - start}')



            print(len(graph._graph.nodes))
            print(len(domain._grid.keys()))

            last_domain = domain
            i.view(last_domain)
            start = time()
            domain = domain.split()
            print(f'Split domain {time() - start}')
            iteration += 1

        return last_domain

File: ComputerSimulation/Task1__CR_set_localisation_CSharp/Task1__CR_set_localisation_CSharp/CR_Localizator.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;
using System.Linq;
using Task1__CR_set_localisation_CSharp.Geometry;
using Task1__CR_set_localisation_CSharp.Graph;

namespace Task1__CR_set_localisation_CSharp
{
    class CR_Localizator
    {
        
        private int _iterationsMaxCount;
        private SymbolicImageGraph _graph;
        private Stopwatch _stopwatch;

        public CR_Localizator(int iterationsMaxCount)
        {
            _iterationsMaxCount = iterationsMaxCount;
        }
        
        public void Solve(Homemorphism f, Domain domain)
        {
            int iterationNumber = 0;

            while (iterationNumber < _iterationsMaxCount)
            {
                Console.WriteLine($"===Iteration {iterationNumber}===");

                ConstructGraph(f, domain);

                DeleteNonReturnableNodes();

                domain.Split();
            }
        }

        private void ConstructGraph(Homemorphism f, Domain domain)
        {
            _stopwatch.Start();
            _graph = new SymbolicImageGraph(f, domain);
            _stopwatch.Stop();

            Console.WriteLine($"Construct Symbolic Image Graph  {_stopwatch.ElapsedMilliseconds / 100.0}");
        }

        private void DeleteNonReturnableNodes()
        {
            _stopwatch.Start();
            _graph.DeleteNonReturnableNodes(out bool successfullyDeleted);
            _stopwatch.Stop();
            if (successfullyDeleted == false)
            {
                throw new NotImplementedException();
            }

            Console.WriteLine($"Delete non-returnable vertexes {_stopwatch.ElapsedMilliseconds}");
        }
    }
}
